doc/adr/0013-download-for-incremental-snapshots.md

```diff
@@ -0,0 +1,91 @@
+# 13. Download for incremental snapshots
+
+Date: 2023-04-18
+
+## Status
+
+Proposed
+
+## Context
+
+Currently, the nodes joining the network are required to download the latest state of the network. However, in case of an extended absence, the node needs to have all incremental snapshots present on disk to determine the latest state of the network. This change requires a modification to the download process, as nodes need to fetch all missing incremental snapshots.
+
+## Decision
+
+After careful consideration of the current network state and the challenges of ensuring consistent and accurate data across all nodes, we have decided to modify the download process for joining nodes to fetch all missing incremental snapshots. This decision is based on the understanding that nodes need to have a complete and validated chain of incremental snapshots to determine the state of the network, especially when they have been absent for an extended period.
+
+The modified download process involves several steps to ensure that nodes have a complete and validated chain of incremental snapshots. The first step is to determine the starting point by querying a peer node for the latest snapshot hash. 
+
+```mermaid
+flowchart TD
+  subgraph Determine starting point
+    A[Get latest snapshot from the Joinee]
+  end
+  subgraph Download process
+	B[Get snapshot with hash X=aaaa]
+  end
+  A --> |latest hash: aaaa| B
+```
+
+The node then checks if it has all the incremental snapshots needed to reach the starting point, and if not, it beings downloading missing snapshots recursively from the latest snapshot to the genesis. All validated snapshots are stored in permanent storage, which contains a correct and validated chain of incremental snapshots. Temporary storage may contain potentially invalid or conflicting snapshots.
+
+Once all the incremental snapshots have been downloaded, the node performs chain validation by applying subsequent incremental snapshots from the genesis to the latest snapshot to build the network state and validate the chain. If the latest snapshot on the network has been updated and the batch is not small enough (determined by the `k` parameter), the node repeats the download process for the next batch of snapshots. Otherwise, it can register for consensus.
```

Comment:
Yes (from the last full snapshot, so it's a genesis for each new network but testnet and mainnet). We know it's not ideal, but to persist the local state, you would have to persist the state for each incremental snapshot, which is no different than having full global snapshots. Node can persist for example the latest state on disk but it doesn't solve the issue of missing or misaligned data.

Reply:
Do we have an estimate on the time this takes? This probably won't cause a severe issue for some time, but we'd want to investigate adding persistence of the state information at some long delay if this becomes too severe. (I believe it's on the order of a day for ETH for example, the re-validation cutoff for 'most' node setups (it can be triggered from genesis though))

Reply:
What is meant by `permanent storage` here?

```
All validated snapshots are stored in permanent storage, which contains a correct and validated chain of incremental snapshots.
```

Can that `permanent storage` be leveraged in this case?

Reply:
What do you mean exactly @cngo-github? Permanent storage is where all validated snapshots are (currently it's a directory on disk with path provided through the CLI parameter, although we may have a broken chain there as for now). We want to make sure we store only the correct chain there.

Reply:
@ryle-goehausen we do not have a time estimation. We need to somehow implement this algorithm to verify how much it takes. I believe that most of the time would be consumed by the network activity (download).

Reply:
@marcinwadon , so download will download from the latest snapshot in the permanent storage to the latest snapshot?

Reply:
Yes, so the happy path is that the node will fetch only the snapshots created during absence (it happens when you restart the node or the node drops out). Worse path would be to remove part of validated state if it comes from another fork and redownload also part of the validated state (it happens during restarts as some nodes are still in the older fork creating snapshots, so they need to rewrite that part). The worst case is when you join as a new node and you have no snapshots so you need to sync the whole history.

Reply:
Yes, unless the node was forked before leaving the network and now joins to a different fork (let's assume "main" fork) -> then it needs to invalidate some of the snapshots that it has in permanent storage. In that case the node will download further until the first correct snapshot is found in permanent storage. Also the direction is different. It will actually download backwards from latest snapshot in the network to the first snapshot from the fork we are joining encountered in the permanent storage.